src/HybridSolver.cpp

```diff
@@ -20,673 +20,275 @@
+        DenseMatrix& AinvDMinvCT_i(AinvDMinvCT_[agg]);
+        DenseMatrix& Ainv_i(Ainv_[agg]);
+        DenseMatrix& hybrid_elem(hybrid_elem_[agg]);
 
-        for (int i = 0; i < nlocal_multiplier; ++i)
-        {
-            const int edgedof_global_id = j_multiplier_edgedof[local_multiplier[i]];
-            const int edgedof_local_id = edgedof_global_to_local_map[edgedof_global_id];
-            Cloc_j[i] = edgedof_local_id;
-            CMinvCT(i) = 1. / M_diag(edgedof_local_id);
-
-            if (edgedof_IsOwned_.RowSize(edgedof_global_id) &&
-                edge_marker[edgedof_global_id])
-            {
-                edge_marker[edgedof_global_id] = false;
-                MinvCT_i.Elem(edgedof_local_id, i) = 1. / M_diag(edgedof_local_id);
-                Cloc_data[i] = 1.;
-            }
-            else
-            {
-                MinvCT_i.Elem(edgedof_local_id, i) = -1. / M_diag(edgedof_local_id);
-                Cloc_data[i] = -1.;
-            }
-        }
-
-        mfem::SparseMatrix Cloc(Cloc_i, Cloc_j, Cloc_data,
-                                nlocal_multiplier, nlocal_edgedof);
-
-        for (int i = 0; i < nlocal_edgedof; ++i)
-            edgedof_global_to_local_map[local_edgedof[i]] = -1;
+        AinvDMinvCT_i.SetSize(nlocal_vertexdof, nlocal_multiplier);
+        hybrid_elem.SetSize(nlocal_multiplier, nlocal_multiplier);
+        Aloc.SetSize(nlocal_vertexdof, nlocal_vertexdof);
+        DMinvCT.SetSize(nlocal_vertexdof, nlocal_multiplier);
 
-        MinvDT_i.InvLeftScaling(M_diag);
+        MinvDT_i.SetSize(Minv.Cols(), Dloc.Rows());
+        MinvCT_i.SetSize(Minv.Cols(), Cloc.Rows());
 
-        tmpHybrid_el.SetSize(nlocal_multiplier, nlocal_multiplier);
+        Dloc.MultCT(Minv, MinvDT_i);
+        Cloc.MultCT(Minv, MinvCT_i);
 
-        Aloc.SetSize(nlocal_vertexdof, nlocal_vertexdof);
+        Cloc.Mult(MinvCT_i, hybrid_elem);
+        Dloc.Mult(MinvCT_i, DMinvCT);
+        Dloc.Mult(MinvDT_i, Aloc);
 
-        if (Dloc.Height() > 0 && Dloc.Width() > 0)
+        if (use_w_)
         {
-            mfem::Mult(Dloc, MinvDT_i, Aloc);
-        }
-        else
-        {
-            Aloc = 0.0;
+            auto Wloc_tmp = mgl.LocalW().GetSubMatrix(local_vertexdof, local_vertexdof, edge_map);
+            Wloc_tmp.ToDense(Wloc);
+
+            Aloc -= Wloc;
         }
 
-        // Compute CMinvDT  = Cloc * MinvDT
-        CMinvDT.SetSize(nlocal_multiplier);
-        column_in.SetDataAndSize(MinvDT_i.Data(), nlocal_multiplier);
-        Cloc.Mult(column_in, CMinvDT);
+        Aloc.Invert(Ainv_i);
 
-        if (W_)
-        {
-            auto Wloc = ExtractRowAndColumns(*W_, local_vertexdof, local_vertexdof,
-                                             edgedof_global_to_local_map);
-            mfem::DenseMatrix tmpW;
-            Full(Wloc, tmpW);
+        Ainv_i.Mult(DMinvCT, AinvDMinvCT_i);
 
-            Aloc -= tmpW;
+        if (DMinvCT.Rows() > 0 && DMinvCT.Rows() > 0)
```

Comment:
One of them should be `Cols()`.  

On an unrelated issue, `DMinvCT.Cols() == 0` when number of local multipliers is 0, which happens when there is an isolated vertex. But I cannot think of any case when `DMinvCT.Rows() == 0` (number of local vertex dofs is 0). I feel like this should not be happening, so an assert maybe more appropriate. Am I missing some edge case?